language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToFunctionCodeAction.java

```diff
@@ -0,0 +1,593 @@
+                }
+                argsForExtractFunction.add(String.format("%s %s", possibleType.get(), symbol.getName().get()));
+            } else if (symbol.kind() == SymbolKind.PARAMETER) {
+                Optional<String> possibleType = CodeActionUtil
+                        .getPossibleType(((ParameterSymbol) symbol).typeDescriptor(), new ArrayList<>(), context);
+                if (possibleType.isEmpty()) {
+                    return Optional.empty();
+                }
+                argsForExtractFunction.add(String.format("%s %s", possibleType.get(), symbol.getName().get()));
+            }
+        }
+
+        ArgLists argLists = new ArgLists();
+        argLists.extractFunctionArgs = argsForExtractFunction;
+        argLists.replaceFunctionCallArgs = argsForReplaceFunctionCall;
+
+        return Optional.of(argLists);
+    }
+
+    private Optional<Node> findEnclosingModulePartNode(NonTerminalNode node) {
+        Node reference = node;
+
+        while (reference != null && reference.parent() != null) {
+            if (reference.parent().kind() == SyntaxKind.MODULE_PART) {
+                return Optional.of(reference);
+            }
+            reference = reference.parent();
+        }
+
+        return Optional.empty();
+    }
+
+    private boolean isWithinFunctionDefinitionNode(NonTerminalNode node) {
+        NonTerminalNode reference = node;
+
+        while (reference != null && reference.parent() != null) {
+            SyntaxKind parentKind = reference.parent().kind();
+            if (parentKind == SyntaxKind.FUNCTION_DEFINITION || parentKind == SyntaxKind.RESOURCE_ACCESSOR_DEFINITION
+                    || parentKind == SyntaxKind.OBJECT_METHOD_DEFINITION) {
+                return true;
+            }
+            reference = reference.parent();
+        }
+
+        return false;
+    }
+
+    private boolean addNewLineAtEnd(Node enclosingNode) {
+        for (Node node : enclosingNode.parent().children()) {
+            if (node.lineRange().startLine().line() == enclosingNode.lineRange().endLine().line() + 1) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /*
+    * With the fix of #37454, we can remove this function
+    * */
+    @Deprecated(forRemoval = true)
+    private static boolean isExpressionNode(NonTerminalNode node) {
+        return (node.kind().compareTo(SyntaxKind.BINARY_EXPRESSION) >= 0
+                    && node.kind().compareTo(SyntaxKind.TYPE_DESC) < 0);
+    }
+
+    private boolean isExpressionExtractable(CodeActionContext context, RangeBasedPositionDetails positionDetails) {
+        NonTerminalNode matchedCodeActionNode = positionDetails.matchedCodeActionNode();
+        Optional<Node> enclosingModulePartNode = findEnclosingModulePartNode(matchedCodeActionNode);
+        boolean isRangeExtractable = true;
+
+        List<SyntaxKind> unSupportedModuleLevelSyntaxKinds = List.of(SyntaxKind.CONST_DECLARATION,
```

Comment:
Let's make this a constant of the class and make it a Set so that contains method will be faster